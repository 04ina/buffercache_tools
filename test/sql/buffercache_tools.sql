--
-- Preparing
--

SET allow_in_place_tablespaces = true;
CREATE TABLESPACE test_tablespace LOCATION '';
RESET allow_in_place_tablespaces;

CREATE DATABASE test_database_1;
CREATE DATABASE test_database_2;

\c test_database_1 \\
CREATE EXTENSION buffercache_tools;

CREATE TABLE test_table_1(col integer) 
    WITH (autovacuum_enabled = off)
    TABLESPACE pg_default;
INSERT INTO test_table_1 
    SELECT 1 FROM generate_series(1,1000); 

CREATE VIEW tt_1 AS 
    SELECT blocknumber, dirty, pinning, fork 
        FROM pg_show_relation_buffers('test_table_1') 
        ORDER BY fork, blocknumber; 

CREATE TABLE test_table_2(col integer) 
    WITH (autovacuum_enabled = off)
    TABLESPACE test_tablespace;
INSERT INTO test_table_2 
    SELECT 1 FROM generate_series(1,1000); 

CREATE VIEW tt_2 AS 
    SELECT blocknumber, dirty, pinning, fork 
        FROM pg_show_relation_buffers('test_table_2') 
        ORDER BY fork, blocknumber; 

\c test_database_2 \\
CREATE EXTENSION buffercache_tools;

CREATE TABLE test_table_3(col integer) 
    WITH (autovacuum_enabled = off)
    TABLESPACE pg_default;
INSERT INTO test_table_3 
    SELECT 1 FROM generate_series(1,1000); 

CREATE VIEW tt_3 AS 
    SELECT blocknumber, dirty, pinning, fork 
        FROM pg_show_relation_buffers('test_table_3') 
        ORDER BY fork, blocknumber; 
VACUUM;
CHECKPOINT;

\c test_database_1 \\

VACUUM;
CHECKPOINT;

-- 
-- Check pg_mark_buffer_dirty() and pg_flush_buffer()
--
SELECT * FROM tt_1;
SELECT pg_mark_buffer_dirty(
    ( 
        SELECT buffernumber 
            FROM pg_show_relation_buffers('test_table_1') 
            ORDER BY fork, blocknumber
            LIMIT 1
    )::integer
);
SELECT * FROM tt_1;
SELECT pg_flush_buffer(
    ( 
        SELECT buffernumber 
            FROM pg_show_relation_buffers('test_table_1') 
            ORDER BY fork, blocknumber 
            LIMIT 1
    )::integer
);
SELECT * FROM tt_1;

-- 
-- Check pg_mark_relation_fork_buffers_dirty() and pg_flush_relation_fork_buffers()
--
SELECT * FROM tt_1;
SELECT pg_mark_relation_fork_buffers_dirty('test_table_1', 'main');
SELECT * FROM tt_1;
SELECT pg_flush_relation_fork_buffers('test_table_1', 'main');
SELECT * FROM tt_1;

-- 
-- Check pg_mark_relation_buffers_dirty() and pg_flush_relation_buffers()
--
SELECT * FROM tt_1;
SELECT pg_mark_relation_buffers_dirty('test_table_1');
SELECT * FROM tt_1;
SELECT pg_flush_relation_buffers('test_table_1');
SELECT * FROM tt_1;

-- 
-- Check pg_mark_database_buffers_dirty() and pg_flush_database_buffers()
--
SELECT * FROM tt_1;
SELECT * FROM tt_2;
\c test_database_2 \\
SELECT * FROM tt_3;
\c test_database_1 \\

SELECT pg_mark_database_buffers_dirty(
    (
        SELECT oid 
            FROM pg_database 
            WHERE datname = 'test_database_1'
    )::oid  
);

SELECT * FROM tt_1;
SELECT * FROM tt_2;
\c test_database_2 \\
SELECT * FROM tt_3;
\c test_database_1 \\

SELECT pg_flush_database_buffers(
    (
        SELECT oid 
            FROM pg_database 
            WHERE datname = 'test_database_1'
    )::oid  
);

SELECT * FROM tt_1;
SELECT * FROM tt_2;
\c test_database_2 \\
SELECT * FROM tt_3;
\c test_database_1 \\

-- 
-- Check pg_mark_tablespace_buffers_dirty() and pg_flush_tablespace_buffers()
--
SELECT * FROM tt_1;
SELECT * FROM tt_2;
\c test_database_2 \\
SELECT * FROM tt_3;
\c test_database_1 \\

SELECT pg_mark_tablespace_buffers_dirty(
    (
        SELECT oid 
            FROM pg_tablespace 
            WHERE spcname = 'test_tablespace'
    )::oid  
);

SELECT * FROM tt_1;
SELECT * FROM tt_2;
\c test_database_2 \\
SELECT * FROM tt_3;
\c test_database_1 \\

SELECT pg_flush_tablespace_buffers(
    (
        SELECT oid 
            FROM pg_tablespace 
            WHERE spcname = 'test_tablespace'
    )::oid  
);

SELECT * FROM tt_1;
SELECT * FROM tt_2;
\c test_database_2 \\
SELECT * FROM tt_3;
\c test_database_1 \\

-- 
-- Check pg_read_page_into_buffer()
--
SELECT (
    (SELECT pg_read_page_into_buffer('test_table_1', 'main', 0)) = 
    (SELECT buffernumber FROM pg_show_relation_buffers('test_table_1') 
        WHERE blocknumber = 0 AND fork = 'main')
);

--
-- Ending
--
DROP TABLE test_table_1;
DROP TABLE test_table_2;
\c test_database_2 \\
DROP TABLE test_table_3;

DROP TABLESPACE test_tablespace;
\c template1 \\

DROP DATABASE test_database_1;
DROP DATABASE test_database_2;