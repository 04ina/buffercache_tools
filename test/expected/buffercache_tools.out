--
-- Preparing
--
SET allow_in_place_tablespaces = true;
CREATE TABLESPACE test_tablespace LOCATION '';
RESET allow_in_place_tablespaces;
CREATE DATABASE test_database_1;
CREATE DATABASE test_database_2;
\c test_database_1 \\
CREATE EXTENSION buffercache_tools;
CREATE TABLE test_table_1(col integer) 
    WITH (autovacuum_enabled = off)
    TABLESPACE pg_default;
INSERT INTO test_table_1 
    SELECT 1 FROM generate_series(1,1000); 
CREATE VIEW tt_1 AS 
    SELECT blocknumber, dirty, pinning, fork 
        FROM pg_show_relation_buf('test_table_1') 
        ORDER BY fork, blocknumber; 
CREATE TABLE test_table_2(col integer) 
    WITH (autovacuum_enabled = off)
    TABLESPACE test_tablespace;
INSERT INTO test_table_2 
    SELECT 1 FROM generate_series(1,1000); 
CREATE VIEW tt_2 AS 
    SELECT blocknumber, dirty, pinning, fork 
        FROM pg_show_relation_buf('test_table_2') 
        ORDER BY fork, blocknumber; 
\c test_database_2 \\
CREATE EXTENSION buffercache_tools;
CREATE TABLE test_table_3(col integer) 
    WITH (autovacuum_enabled = off)
    TABLESPACE pg_default;
INSERT INTO test_table_3 
    SELECT 1 FROM generate_series(1,1000); 
CREATE VIEW tt_3 AS 
    SELECT blocknumber, dirty, pinning, fork 
        FROM pg_show_relation_buf('test_table_3') 
        ORDER BY fork, blocknumber; 
VACUUM;
CHECKPOINT;
\c test_database_1 \\
VACUUM;
CHECKPOINT;
-- 
-- Check pg_mark_dirty_buf() and pg_flush_buf()
--
SELECT * FROM tt_1;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

SELECT pg_mark_dirty_buf(
    ( 
        SELECT buffernumber 
            FROM pg_show_relation_buf('test_table_1') 
            ORDER BY fork, blocknumber
            LIMIT 1
    )::integer
);
 pg_mark_dirty_buf 
-------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | t     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

SELECT pg_flush_buf(
    ( 
        SELECT buffernumber 
            FROM pg_show_relation_buf('test_table_1') 
            ORDER BY fork, blocknumber 
            LIMIT 1
    )::integer
);
 pg_flush_buf 
--------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

-- 
-- Check pg_mark_dirty_relation_fork_buf() and pg_flush_relation_fork_buf()
--
SELECT * FROM tt_1;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

SELECT pg_mark_dirty_relation_fork_buf('test_table_1', 'main');
 pg_mark_dirty_relation_fork_buf 
---------------------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | t     |       0 | main
           1 | t     |       0 | main
           2 | t     |       0 | main
           3 | t     |       0 | main
           4 | t     |       0 | main
           0 | f     |       0 | vm
(9 rows)

SELECT pg_flush_relation_fork_buf('test_table_1', 'main');
 pg_flush_relation_fork_buf 
----------------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

-- 
-- Check pg_mark_dirty_relation_buf() and pg_flush_relation_buf()
--
SELECT * FROM tt_1;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

SELECT pg_mark_dirty_relation_buf('test_table_1');
 pg_mark_dirty_relation_buf 
----------------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | t     |       0 | fsm
           1 | t     |       0 | fsm
           2 | t     |       0 | fsm
           0 | t     |       0 | main
           1 | t     |       0 | main
           2 | t     |       0 | main
           3 | t     |       0 | main
           4 | t     |       0 | main
           0 | t     |       0 | vm
(9 rows)

SELECT pg_flush_relation_buf('test_table_1');
 pg_flush_relation_buf 
-----------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

-- 
-- Check pg_mark_dirty_database_buf() and pg_flush_database_buf()
--
SELECT * FROM tt_1;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

SELECT * FROM tt_2;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

\c test_database_2 \\
SELECT * FROM tt_3;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

\c test_database_1 \\
SELECT pg_mark_dirty_database_buf(
    (
        SELECT oid 
            FROM pg_database 
            WHERE datname = 'test_database_1'
    )::oid  
);
 pg_mark_dirty_database_buf 
----------------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | t     |       0 | fsm
           1 | t     |       0 | fsm
           2 | t     |       0 | fsm
           0 | t     |       0 | main
           1 | t     |       0 | main
           2 | t     |       0 | main
           3 | t     |       0 | main
           4 | t     |       0 | main
           0 | t     |       0 | vm
(9 rows)

SELECT * FROM tt_2;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | t     |       0 | fsm
           1 | t     |       0 | fsm
           2 | t     |       0 | fsm
           0 | t     |       0 | main
           1 | t     |       0 | main
           2 | t     |       0 | main
           3 | t     |       0 | main
           4 | t     |       0 | main
           0 | t     |       0 | vm
(9 rows)

\c test_database_2 \\
SELECT * FROM tt_3;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

\c test_database_1 \\
SELECT pg_flush_database_buf(
    (
        SELECT oid 
            FROM pg_database 
            WHERE datname = 'test_database_1'
    )::oid  
);
 pg_flush_database_buf 
-----------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

SELECT * FROM tt_2;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

\c test_database_2 \\
SELECT * FROM tt_3;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

\c test_database_1 \\
-- 
-- Check pg_mark_dirty_tablespace_buf() and pg_flush_tablespace_buf()
--
SELECT * FROM tt_1;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

SELECT * FROM tt_2;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

\c test_database_2 \\
SELECT * FROM tt_3;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

\c test_database_1 \\
SELECT pg_mark_dirty_tablespace_buf(
    (
        SELECT oid 
            FROM pg_tablespace 
            WHERE spcname = 'test_tablespace'
    )::oid  
);
 pg_mark_dirty_tablespace_buf 
------------------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

SELECT * FROM tt_2;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | t     |       0 | fsm
           1 | t     |       0 | fsm
           2 | t     |       0 | fsm
           0 | t     |       0 | main
           1 | t     |       0 | main
           2 | t     |       0 | main
           3 | t     |       0 | main
           4 | t     |       0 | main
           0 | t     |       0 | vm
(9 rows)

\c test_database_2 \\
SELECT * FROM tt_3;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

\c test_database_1 \\
SELECT pg_flush_tablespace_buf(
    (
        SELECT oid 
            FROM pg_tablespace 
            WHERE spcname = 'test_tablespace'
    )::oid  
);
 pg_flush_tablespace_buf 
-------------------------
 t
(1 row)

SELECT * FROM tt_1;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

SELECT * FROM tt_2;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

\c test_database_2 \\
SELECT * FROM tt_3;
 blocknumber | dirty | pinning | fork 
-------------+-------+---------+------
           0 | f     |       0 | fsm
           1 | f     |       0 | fsm
           2 | f     |       0 | fsm
           0 | f     |       0 | main
           1 | f     |       0 | main
           2 | f     |       0 | main
           3 | f     |       0 | main
           4 | f     |       0 | main
           0 | f     |       0 | vm
(9 rows)

\c test_database_1 \\
-- 
-- Check pg_read_page_into_buf()
--
SELECT (
    (SELECT pg_read_page_into_buf('test_table_1', 'main', 0)) = 
    (SELECT buffernumber FROM pg_show_relation_buf('test_table_1') 
        WHERE blocknumber = 0 AND fork = 'main')
);
 ?column? 
----------
 t
(1 row)

--
-- Ending
--
DROP TABLE test_table_1;
DROP TABLE test_table_2;
\c test_database_2 \\
DROP TABLE test_table_3;
DROP TABLESPACE test_tablespace;
\c template1 \\
DROP DATABASE test_database_1;
DROP DATABASE test_database_2;
